### ISSUE TITLE
Fix R6 List Implementation in GlassBR

### ISSUE THREAD

**Issue by samm82** [JacquesCarette/Drasil#902]
Currently, the first three items of this list in Drasil are being implemented by `zip`ping a list of `QuantityDict`s and a list of `RelationConcept`s, while the rest of the list is implemented via a list of `QDefinition`s.

https://github.com/JacquesCarette/Drasil/blob/46967c44b13f1c82253ecd6ffa63e85cf69ee9cc/code/drasil-example/Drasil/GlassBR/Body.hs#L558-L571

https://github.com/JacquesCarette/Drasil/blob/46967c44b13f1c82253ecd6ffa63e85cf69ee9cc/code/drasil-example/Drasil/GlassBR/Body.hs#L250-L252

## TODO
This hasn't been updated in a while - work is being done in updateGlassReqs branch
- [ ] Update implementation to match stable
- [ ] Sort this list in caseStudies

**Comment by samm82** [comment]
@JacquesCarette @szymczdm Before this fix, the label for three "odd" quantities was "T:[name]" but now they're being treated as a `DataDefinition`, so it is now "DD:[name]". This is just as broken as before, as is there is neither a TM nor a DD that corresponds with any of these quantities. Where should the reference point to/should it point anywhere? The only place these values are really defined is in the Table of Symbols.

My commit (9b92122) might also make it harder to point these to somewhere else; since these are in the same list, it is assumed that they can be referenced in the same way, so this might have to be tackled with pattern matching, etc.

**Comment by JacquesCarette** [comment]
Yes, this is just as broken as before.

So who assumed they were theories, and now assumes they are data definitions?  That's really the source of the problem.

The worrisome thing you say: "The only place these values are really defined is in the Table of Symbols." -- do you mean this is a local definition that only occurs there?  I am confused.

**Comment by szymczdm** [comment]
^ Ditto on the confusion.

**Comment by samm82** [comment]
So apparently I missed these:

LR and _P<sub>b</sub>_ are both defined in the Terminology and Definitions section:
![LR and Pb](https://user-images.githubusercontent.com/35857611/43588016-91b290c6-9639-11e8-9a23-d23f94aa2c5d.png)

and q is defined in an instance model:
![q](https://user-images.githubusercontent.com/35857611/43588099-c90e8cf0-9639-11e8-9591-a7753dd14998.png)

The problem is that the only way to sort by symbols is to have them as part of the same list, which requires them to be the same type. However, most of them are DDs, one is an IM, and two are in the Terminology and Definitions section, which leads to the problem we have now with referencing.

**Comment by JacquesCarette** [comment]
Question: what exactly are you trying to 'sort by symbol'?  I understand that these are of different types, and can't be in the same (Haskell) list. Where is the sort-that-is-failing occurring?

Which one is an IM?

I do think that it would make sense for LR and Pb to be DDs.

**Comment by smiths** [comment]
I'm not sure I understand the problem from the Drasil point of view, since I don't see what the symbols are explicitly tied to whether they are DDs, IMs etc.  The symbol represents the concept.  q means demand.  It might also have an equation for calculating it, but that equation could change within a document, over time, or for a different document.

Regardless of my lack of deep understanding of Drasil, we do have flexibility in changing the manual version.  In retrospect, many of the decisions made in the original document were arbitrary.  I agree with @JacquesCarette that LR and Pb don't make sense as DDs, but q could easily be switched to a DD.  It isn't really an instance model in the sense of the IMs in the other documents.

**Comment by smiths** [comment]
My previous comment needs to be corrected.  I was thinking about more trivial variables when I said that Pb and LR do not make sense as DDs.  From my point of view of the current documentation Pb and LR are instanced models, not terminology.  The current documentation has them as instanced models, but they could be data definitions.

One thing that is reasonably certain, Pb and LR should not be primarily defined as terminology.  The terminology section, which doesn't have a very prominent role in other examples, is much more about natural language and support information.  For encoding information on equations we want DDs or IMs.  It is fine if the terminology repeats a bit of the DD information, but the terminology should not be the source of the information.  I don't even think we have infrastructure for attaching equations to terminology.

The GlassBR example is a bit strange.  The distinctions between DD, T, and IM seem arbitrary.  Part of the problem is that the problem itself does not have a refinement from theory to instance model.  This is natural in the other examples, but it doesn't exist here.  The equations are empirical equations, not derived as special instances of a more abstract theory.

If I had it to do over again right now, I would eliminate the theoretical models all together and have two instanced models.  One for Safety Req-Pb and one for Safety Req-LR.  Everything else would be a data definition.  The thinking is that the instance models provide the main functionality of the software (is the glass safe or not) everything else supports this.

Although I don't quite like it as much, since all structure would be lost, I could probably even be convinced to, at least temporarily, make everything a data definition.

The terminology section is definitely not important enough to be causing us the headaches that it is.  The other forms of information are much more meaningful.

**Comment by samm82** [comment]
So just to clarify, the next steps involve making LR and Pb Data Definitions? And should q be shifted from an IM to a DD as well? @smiths @JacquesCarette @szymczdm

**Comment by smiths** [comment]
I only think you should make changes that are need to solve the current problem.  If we make too many changes there could be rippled effects that we don't want to deal with right now.  @JacquesCarette and @szymczdm can let us know what changes are necessary right now.  What I am saying is that from a "Smith template" point of view, I am fine with changes I listed.

**Comment by samm82** [comment]
@smiths Now that I'm thinking about it, it would be better to leave it until `addLabels` gets merged in, as it will probably change how the referencing would be done.

**Comment by JacquesCarette** [comment]
Whether these 3 items get changed to DD or IM does not matter so much to me - all that matters is that the information they encode be "more available" than it is with the current encoding.

Isn't there a problem with waiting on `addLabels` as GlassBR is broken in part due to this issue? [As per my comments on #891 and its current non-fix].

**Comment by samm82** [comment]
I think this specific issue won't necessarily affect `addLabels`, and I assumed that fixing GlassBR was part of `addLabels`. If an example doesn't compile, it shouldn't be merged into master

**Comment by JacquesCarette** [comment]
Yes, fixing GlassBR is definitely part of `addLabels`. And very important to work on asap.

I am just thinking that a correct fix to GlassBR may well involve changing LR and Pb to be DDs !

**Comment by samm82** [comment]
Hmmm... I'll try switching them over in `addLabels` and see what comes of it.

**Comment by samm82** [comment]
Also - all three of them are instance models. I have no idea how I missed this, but I did 😬

**Comment by JacquesCarette** [comment]
Ok, having them as instance models is "good", in the sense that their definitions are available. Now, I think this discussion has drifted -- it might be worth going back to the original issue, and seeing what is still the issue 'here'.

**Comment by samm82** [comment]
In 85b7e76b8, I started switching `ReqChunk` to take an `ItemType` instead of a `Sentence`, as we have both sentences and lists functioning as requirements in GlassBR. I had to make a new separate file for ItemType to prevent import cycles, and this change is having massive ripple effects - is this a good change to be pursuing or not yet? If not, I think I've done as much on this issue as I can. @JacquesCarette @szymczdm

**Comment by JacquesCarette** [comment]
If IMs and DDs are in a single section, then the whole section should be sorted as one. If they are in different sections, each sorted on their own.

I think that having `ReqChunk` take an `ItemType` instead of a `Sentence` is premature. I think that @Mornix 's work will affect this anyways, so we may as well wait.

**Comment by samm82** [comment]
I'll update stable then (so `make` runs clean), then make a new issue and assign @Mornix.
